Generate the LXD client certificate on-demand

sean-jc opened the pull request:
Delay certificate generation until the client attempts to communicate
with with a remote server.

Change the check for the 'first time using LXD' message to key off the
existence of the client's config directory (create the directory after
the help message has been printed).  Invert the check for the existence
of /var/lib/lxd when printing the help message; the message should be
printed if /var/lib/lxd missing, which implies that 'lxd init' has not
been run.

Fixes #1851

Signed-off-by: Sean Christopherson <sean.j.christopherson@intel.com>

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

stgraber commented:
No, we absolutely want to show the "lxd init" message only if /var/lib/lxd exists.

"lxd init" only works if LXD is running, so it does require a "/var/lib/lxd" to exist.

The check is there so that on systems which only have the client tool but not the daemon, the message is skipped.

sean-jc commented:
Ah, that makes sense.

The scenario I encountered was attempting to issue basic client commands without LXD running, e.g. I ran into `error: Get http://unix.socket/1.0: read unix @->/var/lib/lxd/unix.socket: read: connection reset by peer`.  Would it make sense to add a similar warning to detect lack of LXD for first-time, non-remote client usage?

stgraber commented:
Oh, I thought we had code to catch that kind of error, maybe not that particular one...

It'd certainly make sense for our connection code to catch such error, check if the path does exist and if not, print a user-friendly error message instead (so only keeping the raw errors for uncommon errors).

tych0 commented:
On Wed, Jun 01, 2016 at 01:44:32PM -0700, Stéphane Graber wrote:
> Oh, I thought we had code to catch that kind of error, maybe not that particular one...

I think we do, but I think it broke. It's pretty brittle pattern
matching code, and the golang standard library changed how they export
errors. But in any case, we never did catch ECONNRESET.

> It'd certainly make sense for our connection code to catch such error, check if the path does exist and if not, print a user-friendly error message instead (so only keeping the raw errors for uncommon errors).

If you do decide to update it, the old pattern match is in
/lxc/main.go:main, and the new pattern match probably needs to look
something like this:

https://github.com/juju/juju/blob/master/tools/lxdclient/client.go#L313

(note the use of os.SyscallError vs. syscall.Errno)

In fact, we could make that a library function in client.go or
something and export it so everyone else could use it too.

> ---
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/lxc/lxd/pull/2068#issuecomment-223118851

sean-jc commented:
Updated the commit to revert to the correct var directory checking and to update i18n files (file:line comments).  Let me know if the i18n files are not generally updated in normal PRs and I'll revert those files.

I'll also take a look at updating the error handling.

sean-jc commented on client.go:
```diff
@@ -164,27 +165,30 @@ func NewClient(config *Config, remote string) (*Client, error) {
+		keyf := config.ConfigPath("client.key")
+
+		if !shared.PathExists(certf) || !shared.PathExists(keyf) {
+			fmt.Fprintf(os.Stderr, i18n.G("Generating a client certificate. This may take a minute...")+"\n")
```
Given @tych0's [comment](https://github.com/lxc/lxd/pull/2075/#discussion_r65605901) in #2075, it seems like this message belongs somewhere else.  It also raises the question of whether or not NewClient() should be generating the certificate in the first place.  After rereading the current code, I think there's an implication that it's acceptable to call NewClient() without valid certificates.

tych0 replied:
Yes, I think we shouldn't be writing to `pwd` (or anywhere else on the filesystem) in NewClient, since it may be use by other golang binaries who want to manage their configuration differently. Can we do all the cert generation in lxc/* somewhere and then just use NewClientFromInfo?

stgraber commented:
Looks like this branch will need rebasing too

sean-jc commented:
Reworked the code to use an intermediate `newClient` in main.go to intercept calls to lxd.NewClient and generate the certificate as needed.  I'm not a big fan of the code duplication or possible fragility of using lxd.NewClient, though the latter can be mitigated through static analysis test.

stgraber commented:
Hmm, so this approach fixes a lot of the problems but not all of them.

On a system which doesn't yet have a certificate, the following would break:
 - lxc remote add some-host 1.2.3.4
 - lxc launch some-host:some-image blah

As the remote as seen in lxc/* will be local (unix socket), but the image source will be a remote LXD server, the NewClient call will be done by Init() in client.go which doesn't use your wrapper.


So maybe I'm overlooking something, but wouldn't it be massively easier to just remove the certificate and key generation stuff from lxc/main.go and instead just do it when adding a non-local remote in lxc/remote.go?

sean-jc commented:
When is the client cert actually required?  Is the client cert ever needed for a default remotes, e.g. images?  Or is it safe to assume that only user-defined remotes will require client authentication?  I was being paranoid and assuming that the client certificate needed to be generated prior to any communication with a remote.

stgraber commented:
I think our code will fail to establish any TLS connection without a client certificate. In theory we should only ever need it when:
 - The remote is accessed over https
 - The protocol is LXD
 - The remote isn't flagged as public

sean-jc commented:
Is it valid/supported for a user to "add" a remote by overwriting `config.yml` without dropping in `client.crt`?  For example, using Puppet/Chef/Salt/etc... to set `config.yml`, but relying on the LXD client to generate a certificate.  That would bypass on-demand certificate generation done in the the `remote add` code path.

stgraber commented:
I think it's fine not to support that particular case. We expect people to use the command line tools.

sean-jc commented:
Moved client cert generation into remoteCmd.addServer() and updated tests.  I delayed certificate generation until there is a reasonably high probability that an HTTPS server will be added.

stgraber commented:
That looks reasonable, some potential improvements:
 - Don't generate a cert when adding a simplestreams remote
 - Don't generate a cert when adding a public remote (--public)

We'd also need to make sure that lxc doesn't blow up when talking to https remotes (simplestreams or public) before it has a certificate.

This can be tested with the default remotes:
 - lxc image list images:
 - lxc image list ubuntu:

sean-jc commented:
Further restricted certificate generation and added tests.

stgraber commented:
jenkins: test this please
